sdk/azidentity/chained_token_credential_test.go

```diff
@@ -153,6 +153,220 @@ func TestChainedTokenCredential_GetTokenWithUnavailableCredentialInChain(t *test
+		AuthorityHost: AuthorityHost(srv.URL()),
+	})
+	if err != nil {
+		t.Fatalf("Failed to create environment credential: %v", err)
+	}
+	cred, err := NewChainedTokenCredential([]azcore.TokenCredential{secCred, envCred}, nil)
+	if err != nil {
+		t.Fatalf("unexpected error: %v", err)
+	}
+	tk, err := cred.GetToken(context.Background(), policy.TokenRequestOptions{Scopes: []string{scope}})
+	if err != nil {
+		t.Fatalf("Received an error when attempting to get a token but expected none")
+	}
+	if tk.Token != tokenValue {
+		t.Fatalf("Received an incorrect access token")
+	}
+	if tk.ExpiresOn.IsZero() {
+		t.Fatalf("Received an incorrect time in the response")
+	}
+	if cred.successfulCredential == nil {
+		t.Fatalf("The successful credential was not assigned")
+	}
+	if cred.successfulCredential != secCred {
+		t.Fatalf("The successful credential should have been the secret credential")
+	}
+}
+
+/**
+ * Helps count the number of times a credential is called.
+ */
+type TestCountPolicy struct{ count int }
+
+/**
+ * Helps count the number of times a credential is called.
+ */
+func (p *TestCountPolicy) Do(req *policy.Request) (*http.Response, error) {
+	p.count += 1
+	return req.Next()
+}
+
+func TestChainedTokenCredential_RepeatedGetTokenWithSuccessfulCredential(t *testing.T) {
+	err := initEnvironmentVarsForTest()
+	if err != nil {
+		t.Fatalf("Could not set environment variables for testing: %v", err)
+	}
+	srv, close := mock.NewTLSServer()
+	defer close()
+
+	secretCountPolicy := &TestCountPolicy{}
+	environmentCountPolicy := &TestCountPolicy{}
+
+	srv.AppendResponse(mock.WithBody([]byte(accessTokenRespSuccess)))
+	srv.AppendResponse(mock.WithBody([]byte(accessTokenRespSuccess)))
+	options := ClientSecretCredentialOptions{}
+	options.AuthorityHost = AuthorityHost(srv.URL())
+	options.PerCallPolicies = []policy.Policy{secretCountPolicy}
+	options.Transport = srv
+	secCred, err := NewClientSecretCredential(tenantID, clientID, secret, &options)
+	if err != nil {
+		t.Fatalf("Unable to create credential. Received: %v", err)
+	}
+	envCred, err := NewEnvironmentCredential(&EnvironmentCredentialOptions{
+		ClientOptions: azcore.ClientOptions{
+			Transport:       srv,
+			PerCallPolicies: []policy.Policy{environmentCountPolicy},
+		},
+		AuthorityHost: AuthorityHost(srv.URL()),
+	})
+	if err != nil {
+		t.Fatalf("Failed to create environment credential: %v", err)
+	}
+	cred, err := NewChainedTokenCredential([]azcore.TokenCredential{secCred, envCred}, nil)
+	if err != nil {
+		t.Fatalf("unexpected error: %v", err)
+	}
+	tk, err := cred.GetToken(context.Background(), policy.TokenRequestOptions{Scopes: []string{scope}})
+	if err != nil {
+		t.Fatalf("Received an error when attempting to get a token but expected none")
+	}
+	if tk.Token != tokenValue {
+		t.Fatalf("Received an incorrect access token")
+	}
+	if tk.ExpiresOn.IsZero() {
+		t.Fatalf("Received an incorrect time in the response")
+	}
+	if cred.successfulCredential == nil {
```

Comment:
You don't need to test implementation details like this--counting GetToken calls tells you whether ChainedTokenCredential behaved correctly--and doing so makes the test more fragile.

Reply:
I’m confused about what you’re requesting. I like the idea of checking that this property is set, and counting getToken calls. In any case, please review this new approach! https://github.com/Azure/azure-sdk-for-go/pull/16392/commits/20197a94527b5dc38f8d0270df23c0fccff47e68 It should be better than the previous one 🙂

Reply:
What we want to test is the credential's behavior: does it call its sources' GetToken methods correctly? The count of calls to each source's GetToken method is a direct observation of that. The value of `successfulCredential` is part of that behavior's implementation, not an observation of it. Checking the value of `successfulCredential` like this is validating whether the behavior is implemented the way you expect. The implementation isn't important, and we don't want to bake it into the test. So, I suggest not using `successfulCredential` in these tests.

Reply:
Why is the implementation not important? I wonder what scenarios you’re considering. At a minimum it should have the credential that succeeds 🤔

Reply:
If we change the implementation such that `successfulCredential` no longer exists, this test won't compile. If the credential still behaves correctly, is the broken test helpful?

Reply:
> Even then, these tests would be easy to fix.
> 
It's still work. If we remove `successfulCredential` and the credential still behaves correctly, why would you want the test to break?

> In any case, part of the proposal of this PR is to expose the `successfulCredential` to the public. If that’s not part of the goal, we could take a different direction.
>
That's not the goal or the current state (`successfulCredential` isn't exported).

Reply:
oh I see your point! I’ll remove it from these tests.

Reply:
Updated here: https://github.com/Azure/azure-sdk-for-go/pull/16392/commits/a19c2f27dcbdb3efe444e80cdf7bc59829f0a577 let me know if I missed anything!